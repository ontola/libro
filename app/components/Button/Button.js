import './button.scss';
import React, {Component} from 'react';
import classNames from 'classnames';

export default class Button extends Component {
  constructor(props) {
    super(props);
  }

  render() {

    let btnClass = classNames({
      'btn': true,
      'weight': this.props.weight,
      [`${this.props.theme}`]: true
    });

    return (
      <button {...this.props} className={btnClass}>{this.props.children}</button>
    );
  }
}

Button.propTypes = {
  theme: React.PropTypes.oneOf([
    'subtle',
    'pro',
    'con',
    'default'
  ]),
  weight: React.PropTypes.bool,
  children: React.PropTypes.node.isRequired
};

Button.defaultProps = {
  theme: 'default',
  weight: false,
  children: 'Joep'
};
