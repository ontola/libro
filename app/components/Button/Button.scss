@import '../shared/config';

.Button {
  border: 0;
  display: inline-block;
  font-family: $font-family;
  font-size: font-size('medium');
  margin: 0;
  padding: units('x-small', 'large');
  transition: .1s box-shadow, .1s background-color;

  &:hover {
    cursor: pointer;
  }

  // Not sure about this one, since it may hurt accessibility.
  // The annoying blue box only appears if the button has no action / href. - Joep
  &:focus {
    outline: 0;
  }

  &:hover,
  &:focus {
    box-shadow: inset 0 0 0 999px color('transparent', 'mid-dark');
  }

  &:active {
    box-shadow: inset 0 0 0 999px color('transparent', 'dark');
    transition: 0s box-shadow;
  }
}

.Button--has-icon {
  .Button__icon {
    margin-right: 4px;
  }
}

.Button--default {
  background-color: color('pink');
  border-radius: $border-radius-base;
  color: color('white');
}

.Button--subtle {
  background-color: color('grey', 'light');
  border-radius: $border-radius-base;
  color: color('grey', 'dark');

  &:hover,
  &:focus {
    box-shadow: inset 0 0 0 999px color('grey', 'light');
  }

  &:active {
    box-shadow: inset 0 0 0 999px color('grey', 'x-light');
  }
}

.Button--small {
  font-size: font-size('small');
  padding: units('xx-small', 'x-small');
}

.Button--box {
  background-color: color('white');
  border-right: 1px solid color('grey', 'x-light');
  color: color('grey');
  flex-grow: 1;

  &:last-child {
    border-right: 0;
  }

  &.Button--active {
    background-color: color('grey', 'x-light');
  }
}

@each $variant, $color in $button-variants-colors {
  .Button--variant-#{$variant} {
    color: color($color);
  }
}
