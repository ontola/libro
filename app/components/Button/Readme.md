    <div>
      <p>
        <Button>Default Button</Button> <Button small>Click here</Button>
      </p>
      <p>
        <Button theme="subtle">Click here</Button> <Button small theme="subtle">Click here</Button>
      </p>
      <p>
        <Button icon="check">Default Button with icon</Button> <Button icon="check" small>Default Button with icon</Button>
      </p>
      <p>
        <Button theme="box">Click here</Button> <Button theme="box" variant="pro">Click here</Button> <Button theme="box" variant="con">Click here</Button>
      </p>
    </div>
