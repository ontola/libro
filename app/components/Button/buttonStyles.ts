import makeStyles from '@material-ui/styles/makeStyles';

import { LibroTheme, Margin } from '../../themes/themes';

export enum ButtonVariant {
  AsCard = 'asCard',
  Box = 'box',
  CardBig = 'cardBig',
  Default = 'default',
  Omniform = 'omniform',
  Pagination = 'pagination',
  Submit = 'submit',
  Subtle = 'subtle',
  Toggle = 'toggle',
  Transparent = 'transparent',
  Filled = 'filled',
  Outlined = 'outlined',
}

interface ColorProps {
  color?: string;
}

export default makeStyles<LibroTheme, ColorProps>((theme) => ({
  active: {},

  button: {
    '&:active': {
      boxShadow: 'inset 0 0 0 999px rgba(0, 0, 0, .06)',
      outline: 0,
      transition: '0s box-shadow',
    },
    '&:hover': {
      cursor: 'pointer',
    },
    '&:hover, &:focus': {
      boxShadow: 'inset 0 0 0 999px rgba(0, 0, 0, .03)',
    },
    '&:is(a)': {
      alignItems: 'center',
    },
    '&[disabled]': {
      '&:active, &:hover': {
        boxShadow: 'none',
      },
      cursor: 'not-allowed',
      opacity: 0.5,
    },
    '--button-padding-horizontal': '1.3rem',
    '--button-padding-vertical': '.5rem',
    border: 0,
    display: 'inline-block',
    fontFamily: theme.typography.fontFamily,
    fontSize: '1em',
    margin: 0,
    padding: 'var(--button-padding-vertical) var(--button-padding-horizontal)',
    transition: '.1s box-shadow, .1s background-color',
    whiteSpace: 'nowrap',
  },

  [ButtonVariant.AsCard]: {
    '&:hover, &:focus, &:active': {
      backgroundColor: theme.palette.background.paper,
      boxShadow: '0 2px 8px rgba(0, 0, 0, .2)',
      color: theme.palette.grey.xDark,
    },
    backgroundColor: theme.palette.background.paper,
    borderRadius: theme.shape.borderRadius,
    boxShadow: '0 0 25px rgba(0, 0, 0, .06)',
    color: theme.palette.grey.main,
    marginBottom: '1em',
    transition: '.2s',
  },

  [ButtonVariant.Box]: {
    '&$active': {
      backgroundColor: theme.palette.grey.xLight,
    },
    '&:last-child': {
      borderRight: 0,
    },
    alignSelf: 'stretch',
    borderRight: `1px solid ${theme.palette.grey.xLight}`,
    color: theme.palette.grey.main,
    flexGrow: 1,
  },

  [ButtonVariant.CardBig]: {},

  [ButtonVariant.Default]: {
    backgroundColor: 'var(--accent-background-color, rgb(208, 2, 91))',
    borderRadius: theme.shape.borderRadius,
    color: 'var(--accent-color)',
  },

  [ButtonVariant.Submit]: {
    //@ts-ignore
    backgroundColor: theme.palette.green.main,
    borderBottomLeftRadius: 0,
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
    //@ts-ignore
    color: theme.palette.common.white,
  },

  [ButtonVariant.Subtle]: {
    backgroundColor: theme.palette.grey.xxLight,
    border: `1px solid ${theme.palette.grey.xLight}`,
    borderRadius: theme.shape.borderRadius,
    color: theme.palette.grey.midDark,
  },

  [ButtonVariant.Transparent]: {
    '&$active': {
      '& $label': {
        boxShadow: `inset 0 -2px 0 0 ${theme.palette.grey.main}`,
        fontWeight: theme.typography.fontWeightBold,
      },
      backgroundColor: theme.palette.background.default,
    },
    '&:hover, &:focus': {
      color: theme.palette.grey.dark,
    },
    borderRadius: theme.shape.borderRadius,
    color: theme.palette.grey.main,
  },

  [ButtonVariant.Omniform]: {
    '&$active': {
      '& $label': {
        boxShadow: 'unset',
        fontWeight: theme.typography.fontWeightBold,
      },
      color: theme.palette.grey.midDark,
    },
    '&::after': {
      backgroundColor: 'rgba(0,0,0,0)',
      borderBottomLeftRadius: theme.shape.borderRadius,
      borderBottomRightRadius: theme.shape.borderRadius,
      content: '""',
      height: '6px',
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,
    },
    '&:hover, &:focus': {
      color: theme.palette.grey.dark,
    },
    borderRadius: theme.shape.borderRadius,
    color: theme.palette.grey.midLight,
    overflow: 'hidden',
    paddingLeft: '.8em',
    paddingRight: '.8em',
    position: 'relative',
  },
  [ButtonVariant.Outlined]: {
    '& $label': {
      marginTop: '-4px',
    },
    '&:hover, &:focus': {
      backgroundColor: ({ color }) => color,
      borderColor: ({ color }) => color,
      color: ({ color }) => theme.palette.getContrastText(color ?? theme.palette.primary.main),
    },
    alignItems: 'center',
    border: '1px solid currentColor',
    borderRadius: theme.shape.borderRadius,
    color: ({ color }) => color,
    display: 'flex',
    gap: theme.spacing(Margin.Small),
    padding: theme.spacing(Margin.Medium),
    transition: 'background-color 100ms, color 100ms, border-color 100ms',
  },
  [ButtonVariant.Filled]: {
    '& $label': {
      marginTop: '-4px',
    },
    backgroundColor: ({ color }) => color ?? theme.palette.primary.main,
    borderRadius: theme.shape.borderRadius,
    color: ({ color }) => theme.palette.getContrastText(color ?? theme.palette.primary.main),
    display: 'flex',
    gap: theme.spacing(Margin.Small),
    padding: theme.spacing(Margin.Medium),
  },
  [ButtonVariant.Pagination]: {
    '&$active': {
      '& $label': {
        boxShadow: `inset 0 -2px 0 0 ${theme.palette.grey.main}`,
        fontWeight: theme.typography.fontWeightBold,
      },
      backgroundColor: 'unset',
      color: theme.palette.text.primary,
    },
    '&:hover, &:focus': {
      color: theme.palette.grey.dark,
    },
    '&[disabled]': {
      '& $fa': {
        color: theme.palette.grey.light,
      },
      color: theme.palette.grey.light,
      cursor: 'initial',
      fontWeight: theme.typography.fontWeightBold,
      opacity: 1,
    },
    color: theme.palette.grey.main,
  },

  buttonPlain: {
    '&:hover, &:focus, &:active': {
      boxShadow: 'none',
    },
    '--button-padding-horizontal': 0,
    '--button-padding-vertical': 0,
    alignSelf: 'initial',
    background: 'none',
    color: 'inherit',
  },

  cardFloat: {
    '&:active': {
      color: theme.palette.grey.xDark,
    },
    '&:hover': {
      color: theme.palette.grey.midDark,
    },
    color: theme.palette.grey.xxLightForegroundLarge,
    fontWeight: theme.typography.fontWeightBold,
  },

  cardList: {
    '&:active': {
      color: theme.palette.grey.xDark,
    },
    '&:hover': {
      color: theme.palette.grey.midDark,
    },
    color: theme.palette.grey.xxLightForegroundLarge,
    fontSize: '.8rem',
    fontWeight: theme.typography.fontWeightBold,
    marginBottom: '.7rem',
  },

  centered: {
    '& $label': {
      textAlign: 'center',
      width: '100%',
    },
  },

  corner: {
    '--button-padding-horizontal': '.4rem',
    '--button-padding-vertical': '.4rem',
    lineHeight: '1em',
    position: 'absolute',
    right: 0,
    top: 0,
    zIndex: 1,
  },

  edgeEnd: {
    marginLeft: 'var(--button-padding-horizontal)',
  },

  edgeStart: {
    marginLeft: 'calc(var(--button-padding-horizontal) * -1)',
  },

  fa: {},

  grow: {
    flexGrow: 1,
  },

  hasChildren: {},

  hasIcon: {
    '& $label': {
      marginLeft: '4px',
    },
    '&$hasChildren': {
      '& $label': {
        marginLeft: '4px',
      },
    },
  },

  icon: {},

  label: {},

  margins: {
    margin: '0 .5rem',
  },

  narrow: {
    '--button-padding-horizontal': '.75rem',
    '--button-padding-vertical': '.5rem',
  },

  small: {
    '--button-padding-horizontal': '.5rem',
    '--button-padding-vertical': '.3rem',
    fontSize: '.85em',
  },

  snackbar: {
    '&:active': {
      //@ts-ignore
      background: theme.palette.transparent.light,
    },
    '&:hover': {
      //@ts-ignore
      background: theme.palette.transparent.xLight,
    },
    '--button-padding-horizontal': '.75rem',
    '--button-padding-vertical': '.75rem',
    //@ts-ignore
    color: theme.palette.red.light,
    fontSize: '.8em',
    fontWeight: theme.typography.fontWeightBold,
    textTransform: 'uppercase',
  },

  stretched: {},

  [ButtonVariant.Toggle]: {
    '& $icon': {
      fontSize: '1.2rem',
    },
    '& $label': {
      boxShadow: 'inset 0 -2px 0 0 rgba(0,0,0,0%)',
      padding: '.3em 0',
      transition: 'box-shadow 200ms',
    },
    '&$active': {
      '& $label': {
        fontWeight: theme.typography.fontWeightBold,
      },
      '& $label, & $icon': {
        color: 'var(--buttonText)',
      },
      backgroundColor: 'var(--buttonBackground)',
    },
    '&$cardFloat': {
      '--button-padding-horizontal': '.5rem',
      '--button-padding-vertical': 0,
    },
    '&$stretched': {
      '--button-padding-horizontal': 0,
      '--button-padding-vertical': '.75rem',
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center',
      transition: 'box-shadow 200ms',
    },
    '&:active': {
      backgroundColor: `${theme.palette.grey.xLight} !important`,
    },
    '&:hover, &:focus': {
      '& $label, & $icon': {
        color: 'var(--buttonText)',
      },
      backgroundColor: 'var(--buttonBackground)',
    },
    '--buttonBackground': ({ color }) => color ? color : theme.palette.grey.main,
    '--buttonText': ({ color }) => color ? theme.palette.getContrastText(color) : theme.palette.common.white,
    alignItems: 'center',
    border: '1px solid currentColor !important',
    borderRadius: theme.shape.borderRadius,
    color: ({ color }) => color ? color : theme.palette.grey.main,
    display: 'flex',
    fontWeight: theme.typography.fontWeightMedium,
    justifyContent: 'center',
  },
}));
