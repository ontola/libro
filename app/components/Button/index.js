import './button.scss';
import React, { PropTypes } from 'react';
import classNames from 'classnames';

function Button({ weight, theme, children }) {
  let btnClass = classNames({
    btn: true,
    'btn--weight': weight,
    [`btn--${theme}`]: true,
  });

  return (
    <button className={btnClass}>{children}</button>
  );
}

Button.propTypes = {
  theme: PropTypes.oneOf([
    'box',
    'subtle',
    'default',
  ]),
  weight: PropTypes.bool,
  children: PropTypes.node.isRequired,
};


Button.defaultProps = {
  theme: 'default',
  weight: false,
  children: 'Joep',
};

export default Button;
