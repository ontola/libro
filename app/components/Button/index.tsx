import useMediaQuery from '@material-ui/core/useMediaQuery';
import { useTheme } from '@material-ui/styles';
import { NamedNode } from '@ontologies/core';
import clsx from 'clsx';
import { useLRS } from 'link-redux';
import React, { MouseEvent } from 'react';
import FontAwesome from 'react-fontawesome';

import { normalizeFontAwesomeIRI } from '../../helpers/iris';
import { LibroTheme } from '../../themes/themes';
import BlurButton from '../BlurButton';
import LinkDuo from '../LinkDuo';

import useStyles, {
  ButtonTheme,
  ButtonVariant,
} from './buttonStyles';

export { ButtonTheme, ButtonVariant } from './buttonStyles';

export const ButtonLabelIdentifierClass = 'Button__label';
const ButtonIdentifierClass = 'Button';
const ButtonIconIdentifierClass = 'Button__icon';

export interface ButtonProps {
  action?: NamedNode,
  /** Should be true when the button is toggleable and toggled. */
  active?: boolean;
  /** Additional aria label */
  ariaLabel?: string;
  list?: boolean;
  cardFloat?: boolean;
  centered?: boolean;
  /** Label of the button */
  children?: React.ReactNode;
  /** Should be avoided. Try to use the 'theme' prop or wrap it in some other element for styling */
  /** and use the 'plain' prop. */
  className?: string;
  /** Displays button on top right position of relative parent */
  corner?: boolean;
  disabled?: boolean;
  /** FontAwesome icon string. */
  endIcon?: React.ReactNode;
  /** Whether the button should fill it's container */
  grow?: boolean;
  /** If the button is actually just a link */
  href?: string;
  /** FontAwesome icon string. */
  icon?: string;
  /** Set to true if the button should indicate that something is happening. Renders spinner. */
  loading?: boolean;
  /** Give some margins for inline usage */
  margins?: boolean;
  narrow?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  /** Removes all styling. */
  plain?: boolean;
  small?: boolean;
  stretch?: boolean;
  /** Removes all styling. */
  theme?: ButtonTheme;
  /** Title html tag. */
  title?: string;
  type?: string;
  variant?: ButtonVariant;
}

const defaultProps = {
  corner: false,
  grow: false,
  narrow: false,
  plain: false,
  small: false,
  theme: ButtonTheme.Default,
  type: 'button',
};

/**
 * Used for any action in the app. Handles touch events and blurs after handling the onClick.
 * @returns {component} Component
 */
const Button = React.forwardRef<HTMLButtonElement | HTMLAnchorElement, ButtonProps>(({
  active,
  action,
  ariaLabel,
  cardFloat,
  list,
  centered,
  children,
  className,
  corner,
  disabled,
  endIcon,
  grow,
  icon,
  loading,
  margins,
  onClick,
  small,
  stretch,
  narrow,
  plain,
  theme,
  title,
  type,
  href,
  variant,
  ...otherProps
}, ref) => {
  const lrs = useLRS();
  const muiTheme = useTheme<LibroTheme>();
  const isMobile = useMediaQuery(muiTheme.breakpoints.down('sm'));
  const classes = useStyles();
  const handleAction = React.useCallback((e: MouseEvent) => {
    e.preventDefault();

    return lrs.exec(action!);
  }, [action]);
  const handleClick = onClick || (action ? handleAction : undefined);
  const btnClass = clsx({
    [ButtonIdentifierClass]: true,
    [classes.button]: true,
    [classes.buttonPlain]: plain,
    [classes.active]: active,
    [classes.centered]: centered,
    [classes.corner]: corner,
    [classes.grow]: grow,
    [classes.hasChildren]: children,
    [classes.hasIcon]: icon,
    [classes.margins]: margins,
    [classes.narrow]: narrow,
    [classes.small]: small,
    [classes.cardList]: list,
    [classes.cardFloat]: cardFloat,
    [classes.stretched]: stretch,
    [classes[theme ?? '']]: theme,
    [classes[variant ?? '']]: variant,
    [className || '']: className,
  });

  const buttonLabelClass = clsx([
    ButtonLabelIdentifierClass,
    classes.label,
  ]);

  const buttonIconClass = clsx([
    ButtonIconIdentifierClass,
    classes.icon,
  ]);

  const currentIcon = loading ? 'spinner' : icon && normalizeFontAwesomeIRI(icon);

  const sharedProps = {
    'aria-label': ariaLabel,
    'className': btnClass,
    'disabled': disabled || loading,
    onClick: handleClick,
    title,
    type,
  };

  if (typeof href !== 'undefined') {
    return (
      <LinkDuo
        data-test="Button-link"
        role="button"
        to={href}
        {...sharedProps}
        {...otherProps}
        ref={ref as React.Ref<HTMLAnchorElement>}
      >
        {currentIcon && (
          <FontAwesome
            className={buttonIconClass}
            data-test="Button-icon"
            name={currentIcon}
            spin={loading}
          />
        )}
        <span className={buttonLabelClass}>
          {children}
        </span>
        {endIcon}
      </LinkDuo>
    );
  }

  return (
    <BlurButton
      aria-pressed={active}
      data-test="Button-button"
      ref={ref as React.Ref<HTMLButtonElement>}
      {...sharedProps}
      {...otherProps}
    >
      {currentIcon && (
        <FontAwesome
          className={buttonIconClass}
          data-test="Button-icon"
          data-testid="Button-icon"
          name={currentIcon}
          spin={loading}
        />
      )}
      {children && (!icon || !isMobile) && (
        <span className={buttonLabelClass}>
          {children}
        </span>
      )}
      {endIcon}
    </BlurButton>
  );
});

Button.defaultProps = defaultProps;
Button.displayName = 'Button';

export default Button;
