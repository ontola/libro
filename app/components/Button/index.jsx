import './Button.scss';
import React, { PropTypes } from 'react';
import classNames from 'classnames';
import FontAwesome from 'react-fontawesome';
import { buttonThemes, sides } from 'components/shared/config';

const propTypes = {
  active: PropTypes.bool.isRequired,
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  icon: PropTypes.string,
  small: PropTypes.bool,
  theme: PropTypes.oneOf(buttonThemes),
  variant: PropTypes.oneOf(sides),
};

const defaultProps = {
  small: false,
  theme: 'default',
  active: false,
};

const Button = ({
  active,
  children,
  icon,
  onClick,
  small,
  theme,
  variant,
}) => {
  const btnClass = classNames({
    Button: true,
    'Button--has-icon': icon,
    'Button--small': small,
    [`Button--${theme}`]: theme,
    [`Button--variant-${variant}`]: variant,
    'Button--active': active,
  });

  return (
    <button
      onClick={onClick}
      className={btnClass}
      role="button"
      type="button"
    >
      {icon && <FontAwesome className="Button__icon" name={icon} />}
      <span className="Button__label">{children}</span>
    </button>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

export default Button;
