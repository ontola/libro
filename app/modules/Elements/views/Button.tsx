import * as schema from '@ontologies/schema';
import {
  FC,
  Resource,
  register,
  useIds,
  useStrings,
} from 'link-redux';
import React from 'react';
import FontAwesome from 'react-fontawesome';

import { allTopologies } from '../../../topologies';
import argu from '../../Argu/ontology/argu';
import Button, { ButtonVariant } from '../../Common/components/Button';
import elements from '../ontology/elements';

const ElementsButton: FC = (props) => {
  const [text] = useStrings(schema.text);
  const [trackingId] = useStrings(argu.trackingId);
  const [variant] = useStrings(elements.variant);
  const [color] = useStrings(elements.color);
  const [icon] = useStrings(schema.image);
  const [iconPosition] = useStrings(elements.iconPosition);
  const [href] = useIds(elements.href);
  const [children] = useIds(elements.children);

  const endIcon = React.useMemo(() => iconPosition !== 'end' ?
    null : (
      <FontAwesome
        data-testid="button-end-icon"
        name={icon}
      />
    ), [iconPosition, icon]);

  return (
    <Button
      color={color}
      endIcon={endIcon}
      href={href?.value}
      icon={iconPosition === 'start' ? icon : undefined}
      variant={variant as ButtonVariant}
      {...(trackingId ? { id: trackingId } : {})}
      ref={props.innerRef}
    >
      {props.children ?? (children ? <Resource subject={children} /> : text)}
    </Button>
  );
};

ElementsButton.type = elements.Button;
ElementsButton.topology = allTopologies;

export default register(ElementsButton);
