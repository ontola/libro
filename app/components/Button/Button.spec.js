import Button from './index';

const clickHandler = jest.fn();
const link = 'http://argu.co/resource';

argUnit(Button, () => {
  set('onClick', () => clickHandler);

  it('should render', () => {
    expect(clickHandler).not.toHaveBeenCalled();
    expect(subject).toMatchSnapshot();
  });

  it('should be clickable', () => {
    expect(subject.simulate('click'));
    expect(clickHandler).toHaveBeenCalled();
  });

  it('should not be disabled', () => {
    expect(subject.find(marker('button'))).toHaveProp('disabled', undefined);
  });

  it('should render a button', () => {
    expect(subject.find(marker('link'))).not.toBePresent();
    expect(subject.find(marker('button'))).toBePresent();
  });

  describe('with custom class', () => {
    set('className', () => 'custom');

    it('should render subtle', () => {
      expect(subject).toHaveClassName('custom');
    });
  });

  describe('with href', () => {
    set('href', () => link);

    it('should render a link', () => {
      expect(subject.find(marker('link'))).toBePresent();
      expect(subject.find(marker('button'))).not.toBePresent();
    });

    it('should point to the link', () => {
      expect(subject.find(marker('link'))).toHaveProp('to', link);
    });

    it('should not be disabled', () => {
      expect(subject.find(marker('link'))).toHaveProp('disabled', undefined);
    });

    describe('with icon', () => {
      set('icon', () => 'lightbulb-o');

      it('should render an icon', () => {
        expect(subject.find(marker('icon'))).toHaveProp('name', 'lightbulb-o');
      });
    });
  });

  describe('with icon', () => {
    set('icon', () => 'lightbulb-o');

    it('should render an icon', () => {
      expect(subject.find(marker('icon'))).toHaveProp('name', 'lightbulb-o');
    });
  });

  describe('with label', () => {
    set('children', () => 'Button label');

    it('should render a label', () => {
      expect(subject.find('.Button__label')).toHaveText('Button label');
    });
  });

  describe('when loading', () => {
    set('loading', () => true);

    it('should be disabled', () => {
      expect(subject.find(marker('button'))).toHaveProp('disabled', true);
    });
  });

  describe('with small', () => {
    set('small', () => true);

    it('should render small', () => {
      expect(subject).toHaveClassName('Button--small');
    });
  });

  describe('with subtle', () => {
    set('theme', () => 'subtle');

    it('should render subtle', () => {
      expect(subject).toHaveClassName('Button--subtle');
    });
  });

  describe('with variant', () => {
    set('variant', () => 'pro');

    it('should render subtle', () => {
      expect(subject).toHaveClassName('Button--variant-pro');
    });
  });
});
