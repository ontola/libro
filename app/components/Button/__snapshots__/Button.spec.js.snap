// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render 1`] = `
<button
  className="Button Button--default"
  data-test="Button-button"
  onClick={[Function]}
  type="button"
>
  <span
    className="Button__label"
  />
</button>
`;
